### PR TITLE
Local classes are uncheckable (type tests)

### PR DESCRIPTION
**Pull request by dwijnand**

### CODE REVIEW

**Comment by dwijnand** [comment]
I didn't look into the detail, but it looks like doing it so bluntly breaks:

```
 - dotty.communitybuild.CommunityBuildTestB.http4s
 - dotty.communitybuild.CommunityBuildTestB.scodec
 - dotty.communitybuild.CommunityBuildTestB.fs2
```

**Comment by dwijnand** [comment]
@smarter maybe you're interested to review this, seeing as it's an old one you created.

**Comment by smarter** [comment]
No time to review this unfortunately, but traversing all types of all declarations of all base classes could be pretty expensive.

**Comment by dwijnand** [comment]
It's all types of all declarations of all base classes defined in the method, is hopefully ok.  And I'm all ears for whether there's a better way to do it.

**Review by odersky — CHANGES_REQUESTED** [review]
I also think we should keep searching for a more efficient way to fix this. But I don't have the time right now to do it. So it's probably best to revisit this later.

**Comment by smarter** [comment]
Maybe we should consider all type tests to local classes as unchecked by default, the one case which is safe is when we go from one local class to another local class defined in the same scope:
```scala
def foo = {
  sealed trait A
  class B extends A
  val x: A = new B
  x match { case x: B => x }
}
```
Anything else could end up giving you an instance of a local class coming from a different call to the same method which semantically is a different class.

**Comment by smarter** [comment]
In other words the check would be something like "`(x: A).isinstanceOf[B]` emits a warning if `B` is a local class which is not statically reachable from the scope where `A` is defined"

**Comment by dwijnand** [comment]
> Anything else could end up giving you an instance of a local class coming from a different call to the same method which semantically is a different class.

That seems overly strict.  A local class that doesn't depend on anything in the method surely isn't "unchecked" when used, such as:

```scala
class A
def foo = {
  class B extends A
  val x: A = new B
  x match {
    case x: B =>
    case _ =>
  }
}
```

But I did realise from your comment that I'm not considering singleton types based on any terms in the method, like any parameters.

Going back to performance, the check here sounds big, but it's "only" on the types of the declarations in the local classes.  So if you have a local class with 30 methods, or 10 local classes with 3 methods, it will take the time that takes.  But I think it's fairly limited, as local classes tend to be few and small, in my experience.  In the case of fs2, https://github.com/dotty-staging/fs2/blob/main/core/shared/src/main/scala/fs2/Pull.scala#L824-L1260 I think I counted 8 classes with about 5 methods each, which seems excessive to me.  I'm also not caching this information anywhere, admittingly, so I could look into that if desired.

**Comment by smarter** [comment]
Your example matches the exception to the unchecked warning I gave above.

**Comment by dwijnand** [comment]
Oh.  Then how are you defining "statically reachable", because I'd assumed being defined in a def disqualified it.

**Comment by smarter** [comment]
I meant reachable through only static references from a specific scope (which might be the scope of a def)

**Comment by smarter** [comment]
Oh but actually I misread your example, i didn't see that A was defined outside the def and B inside. Then I assert that we should get an unchecked warning: in general you cannot know if the B you get was created in the current call to the def or in a different call and escaped, and in the latter case it's really a different class B that happens to have the same runtime representation

**Comment by dwijnand** [comment]
Right.  Do we have some verbiage to back up that claim?  Because I think many users consider the fact that classes can be defined locally to be a proximity convenience rather than also implying method-call uniqueness in those class types.  I find it particularly hard to defend that claim because it's impossible to build a true neg test for that match, using my class definitions.

**Comment by odersky** [comment]
Note: `classSymbols` is not the same as `baseClasses`.

**Comment by smarter** [comment]
> classSymbols of Foo | Bar is {Foo, Bar}.

Is it? `classSymbols` calls `parentSymbols` which does https://github.com/lampepfl/dotty/blob/0260d758d89d01ab445c471aa159ca8367d596ba/compiler/src/dotty/tools/dotc/core/Types.scala#L539-L540
(not sure what the spec in question is)

**Comment by odersky** [comment]
Ah sorry, I meant `Bar & Foo`. For `Bar | Foo`, classSymbols is empty. But in any case it has nothing to do with baseClasses.

**Comment by dwijnand** [comment]
Right, sorry, parentSymbols, not baseClasses.

**Review by odersky — COMMENTED** [review]
Otherwise LGTM

File: compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala [hunk 1]
```diff
@@ -58,7 +59,7 @@ object TypeTestsCasts {
       def apply(tp: Type) = tp match {
         case tref: TypeRef if tref.typeSymbol.isPatternBound =>
           WildcardType
-        case AnnotatedType(_, annot) if annot.symbol == defn.UncheckedAnnot =>
```
[the remark this answers is not in the document]
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala:61]: Ah, I see that now: the previous behaviour mapped `T @foo @unchecked @bar` to `? @bar` while my diff made it map to `?`, but at the quadratic cost of each annotation when there isn't an `@unchecked`.  I'll revert.

File: compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala [hunk 2]
```diff
@@ -152,10 +153,13 @@ object TypeTestsCasts {
...
       case tp2: RefinedType     => recur(X, tp2.parent) && TypeComparer.hasMatchingMember(tp2.refinedName, X, tp2)
       case tp2: RecType         => recur(X, tp2.parent)
+      case _
+      if P.classSymbol.isLocal && foundClasses(X).exists(P.classSymbol.isInaccessibleChildOf) => // 8
+        false
       case _                    => true
     })
 
-    val res = recur(X.widen, replaceP(P))
+    val res = X.widenTermRefExpr.hasAnnotation(defn.UncheckedAnnot) || recur(X.widen, replaceP(P))
```
[the remark this answers is not in the document]
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala:162]: This is for whether the scrutinee was annotated `@unchecked`:

```scala
  def test8[T](x: T): T =
    class A(val elem: T)
    val p = prev
    (p: @unchecked) match
      case prev: A => prev.elem
      case _       => prev = new A(x); x
```

**Comment by dwijnand** [comment]
Let me know if you're happy with my answer (assuming, as I do, that CI passes).

**Review by odersky — APPROVED** [review]
Yes, looks all good to me now,

**Pull request merged**